client/alloc_watcher.go

```diff
@@ -0,0 +1,448 @@
+
+// newAllocWatcher creates a prevAllocWatcher appropriate for whether this
+// alloc's previous allocation was local or remote. If this alloc has no
+// previous alloc then a noop implementation is returned.
+func newAllocWatcher(alloc *structs.Allocation, prevAR *AllocRunner, rpc rpcer, config *config.Config, l *log.Logger) prevAllocWatcher {
+	if alloc.PreviousAllocation == "" {
+		// No previous allocation, use noop transitioner
+		return noopPrevAlloc{}
+	}
+
+	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+
+	if prevAR != nil {
+		// Previous allocation is local, use local transitioner
+		return &localPrevAlloc{
+			allocID:      alloc.ID,
+			prevAllocID:  alloc.PreviousAllocation,
+			tasks:        tg.Tasks,
+			sticky:       tg.EphemeralDisk != nil && tg.EphemeralDisk.Sticky,
+			prevAllocDir: prevAR.GetAllocDir(),
+			prevListener: prevAR.GetListener(),
+			prevWaitCh:   prevAR.WaitCh(),
+			prevStatus:   prevAR.Alloc(),
+			logger:       l,
+		}
+	}
+
+	return &remotePrevAlloc{
+		allocID:     alloc.ID,
+		prevAllocID: alloc.PreviousAllocation,
+		tasks:       tg.Tasks,
+		config:      config,
+		migrate:     tg.EphemeralDisk != nil && tg.EphemeralDisk.Migrate,
+		rpc:         rpc,
+		logger:      l,
+	}
+}
+
+// localPrevAlloc is a prevAllocWatcher for previous allocations on the same
+// node as an updated allocation.
+type localPrevAlloc struct {
+	// allocID is the ID of the alloc being blocked
+	allocID string
+
+	// prevAllocID is the ID of the alloc being replaced
+	prevAllocID string
+
+	// tasks on the new alloc
+	tasks []*structs.Task
+
+	// sticky is true if data should be moved
+	sticky bool
+
+	// prevAllocDir is the alloc dir for the previous alloc
+	prevAllocDir *allocdir.AllocDir
+
+	// prevListener allows blocking for updates to the previous alloc
+	prevListener *cstructs.AllocListener
+
+	// prevStatus allows checking if the previous alloc has already
+	// terminated (and therefore won't send updates to the listener)
+	prevStatus terminated
+
+	// prevWaitCh is closed when the previous alloc is GC'd which is a
+	// failsafe against blocking the new alloc forever
+	prevWaitCh <-chan struct{}
+
+	logger *log.Logger
+}
+
+// Wait on a local alloc to become terminal, exit, or the context to be done.
+func (p *localPrevAlloc) Wait(ctx context.Context) error {
+	defer p.prevListener.Close()
+
+	if p.prevStatus.Terminated() {
+		// Fast path - previous alloc already terminated!
+		return nil
+	}
+
+	// Block until previous alloc exits
+	p.logger.Printf("[DEBUG] client: alloc %q waiting for previous alloc %q to terminate", p.allocID, p.prevAllocID)
+	for {
+		select {
+		case prevAlloc, ok := <-p.prevListener.Ch:
+			if !ok || prevAlloc.Terminated() {
+				return nil
+			}
+		case <-p.prevWaitCh:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Migrate from previous local alloc dir to destination alloc dir.
+func (p *localPrevAlloc) Migrate(ctx context.Context, dest *allocdir.AllocDir) error {
```

Comment:
Great question. I think I'm going to change this to return the error from allocdir.Move but always call allocdir.Destroy since it's a cleanup method we always want to call.

Since the caller logs errors from this method and continues nothing will change functionally but it will make this local Migrate method work like the remote Migrate method.